docs/adrs/004-SplitAndAggregate.md

```diff
@@ -125,12 +125,21 @@ Still to be determined (not in scope of this ADR):
 * `+` Simplest solution, quickest turnaround, fixes deadlock issue
 * `-` Addresses only the deadlocking issue, UI still broken for regular runs and batch runs 
 
-### Increase number of slots or Seperate Queue for kbparallels apps without 10 job limit
-* `+` Simple solutions, quick turnarounds, fixes deadlock issue
+### Increase number of slots per user > 10
+* `+` Simple solutions, quick turnarounds, fixes deadlock issue for small numbers of jobs.
+* `-` Doesn't fix deadlock issue as the user can still submit more KBP jobs
 * `-` Addresses only the deadlocking issue, UI still broken for regular runs and batch runs
 * `-` A small amount of users can take over the entire system
 * `-` The calculations done by the apps will interfere with other apps and cause crashes/failures
 
+###  Seperate Queue for kbparallels apps that may or may not have its own limit to running jobs.
+* `+` Simple solutions, quick turnarounds, fixes deadlock issue 
+* `+` Requires minimum changes to ee2 and condor if condor supports this feature
```

Comment:
What changes would it require?

Reply:
ee2 * Make a list of apps get submitted to that queue
condor * somehow make the 2nd queue a consumable resource

Reply:
> Make a list of apps get submitted to that queue

can't you just specify that in the catalog?

> somehow make the 2nd queue a consumable resource

I don't understand what this means - why would this queue be any different than any other queue?

Reply:
You could specify it in the catalog, and then make ee2 add a KBP_LIMIT if it has that QUEUE clientgroup, but if we want those KBP jobs to run in the NJS queue and limit the max of those jobs, EE2 would need to keep a list of the apps it needs to add the KBP_LIMIT to.

As for why making it a consumable resource, is because you don't want to have one user take over the entire KBP queue, since those jobs wouldn't take up slots anymore. If we limit the amount of jobs sent to the KBP queue, we might as well just not make a KBP queue, because we can limit KBP jobs without them needing their own queue.

Reply:
> As for why making it a consumable resource, is because you don't want to have one user take over the entire KBP queue, since those jobs wouldn't take up slots anymore. If we limit the amount of jobs sent to the KBP queue, we might as well just not make a KBP queue, because we can limit KBP jobs without them needing their own queue.

I'm still not quite grokking this - if we have a combined queue and a regular limit of 10 and a KBP limit of, say, 5, that means that a user running a lot of parallel jobs will be running at 50% capacity.

If instead we have a separate queue with a KBP limit of, say, 20, then the user can run at 100% capacity and run more KBP jobs at the same time. Theoretically we could also assign very low resources per KBP job (especially if we restructured the jobs to do most of the work on the regular queue).

Reply:
* Modifying code inside of apps that use KBP is not in scope of this solution
What is a combined queue?

We have 60 cs nodes, which means we can have 60 jobs that take 60 nodes right now.
If we convert 20 nodes to KBP nodes and we had a limit of 20 jobs per user, then 1 user would submit 20 KBP jobs and no one else could run a KBP job. 20 KBP jobs would be set to running. Of those 20, only up to 10 could have a computation job running.  The other 10 would be doing nothing, as the users 10 slots are all taken up.

Reply:
Ok, gotcha, so it makes sense to make KBP_limit <= regular limit.

> If we convert 20 nodes to KBP nodes and we had a limit of 20 jobs per user, then 1 user would submit 20 KBP jobs and no one else could run a KBP job. 

This is where I'm wondering if we can break the 1 node : 1 job rule to allow more KBP jobs to run at the same time.

Reply:
> Modifying code inside of apps that use KBP is not in scope of this solution

Why not?

>  What is a combined queue?

I'm just talking about the case where KBP jobs and regular jobs run in the same queue.

Reply:
Also FYI having more than 1 job do something like combine stuff with DFU or upload things on the same cs node might cause too much disk pressure on these older machines, so thats we changed to a max of 1 job per machine.

Reply:
* This is where I'm wondering if we can break the 1 node : 1 job rule to allow more KBP jobs to run at the same time.
You sure can, but then you need to consider how many CPUS/MEMORY/DISK each KBP job is using and how many can fit in a node. If we do it wrong, then the jobs are all going to be affecting each other, causing more user tickets to come in complaining about their jobs mysteriously crashing. 

* Modifying code inside of apps that use KBP is not in scope of this solution
Well you have to cut off the solution at some point before you make a new one.
It seems very complex to come up with 7 plans of modifying KBP apps that make them not do any computation on the same machines and not affect other jobs, and weigh that against just ripping KBP out of those apps.

KBP and regular jobs already run in the same queue, see the proposal that allows the jobs to run on their own nodes and also fixes deadlocking *LIMIT KBP jobs to a maximum of 5 active jobs per user*